src/19FAF0.c

```diff
@@ -126,12 +172,130 @@ void dispatch_event_player_continue_turn(s32 eventType) {
     }
 }
 
-INCLUDE_ASM(s32, "19FAF0", calc_player_test_enemy);
+s32 calc_player_test_enemy(void) {
+    BattleStatus* battleStatus = &gBattleStatus;
+    Actor* player = battleStatus->playerActor;
+    s32 currentTargetID = battleStatus->currentTargetID;
+    s32 targetPart = battleStatus->currentTargetPart;
+    ActorState* state = &player->state;
+    Actor* target;
+    ActorPart* part;
+
+    battleStatus->currentTargetID2 = battleStatus->currentTargetID;
+    battleStatus->currentTargetPart2 = battleStatus->currentTargetPart;
+
+    target = get_actor(currentTargetID);
+
+
+    if (target == NULL) {
+        return 0;
+    }
+
+    part = get_actor_part(target, targetPart);
+    ASSERT(part != NULL);
+
+    if (part->eventFlags & ACTOR_EVENT_FLAG_ILLUSORY) {
+        return 6;
+    } else if (target->transStatus == STATUS_E || ((part->eventFlags & ACTOR_EVENT_FLAG_800) &&
+               !(battleStatus->currentAttackElement & DAMAGE_TYPE_QUAKE)))
+    {
+        return 6;
+    } else if (target->stoneStatus == STATUS_STONE) {
+        sfx_play_sound_at_position(SOUND_IMMUNE, 0, state->goalPos.x, state->goalPos.y, state->goalPos.z);
+        return 8;
+    } else if ((battleStatus->currentAttackElement & DAMAGE_TYPE_JUMP) && (part->eventFlags & ACTOR_EVENT_FLAG_SPIKY_TOP) &&
+               !player_team_is_ability_active(player, 2))
+    {
+        sfx_play_sound_at_position(SOUND_108, 0, state->goalPos.x, state->goalPos.y, state->goalPos.z);
+        return 4;
+    } else if (!(battleStatus->currentAttackElement & (DAMAGE_TYPE_NO_CONTACT | DAMAGE_TYPE_JUMP)) &&
+               (part->eventFlags & ACTOR_EVENT_FLAG_SPIKY_FRONT) &&
+               (!(battleStatus->currentAttackEventSuppression & 4) &&
+               !player_team_is_ability_active(player, 2)))
+    {
+        sfx_play_sound_at_position(SOUND_108, 0, state->goalPos.x, state->goalPos.y, state->goalPos.z);
+        dispatch_damage_event_player_1(1, EVENT_SPIKE_CONTACT);
+        dispatch_event_actor(target, EVENT_SPIKE_TAUNT);
+        return -1;
+    } else if (player->staticStatus != STATUS_STATIC && target->staticStatus == STATUS_STATIC) {
+        return 7;
+    }
+
+    return 0;
+}
 
 INCLUDE_ASM(s32, "19FAF0", calc_player_damage_enemy);
 
-s32 dispatch_damage_event_player(s32, s32, s32);
-INCLUDE_ASM(s32, "19FAF0", dispatch_damage_event_player);
+s32 dispatch_damage_event_player(s32 damageAmount, s32 event, s32 arg2) {
+    BattleStatus* battleStatus = &gBattleStatus;
+    PlayerData* playerData = &gPlayerData;
+    Actor* player = battleStatus->playerActor;
+    ActorState* state = &player->state;
+    s32 oldHPChangeCounter;
+    s32 flags;
+    s32 dispatchEvent;
+    s32 oldPlayerHP;
+    s32 temp;
+
+    battleStatus->currentAttackDamage = damageAmount;
+
+    temp = (s16)damageAmount; // usage of temp here required to match
```

Comment:
We're sure dispatch_damage_event_player has the right signature?

Reply:
yeah, damageAmount can't be s16 cause it messes up callers